src/eigenvalue.cpp

```diff
@@ -0,0 +1,654 @@
+      simulation::master_fission_bank.clear();
+    } else {
+      simulation::fission_bank.clear();
+    }
+  }
+}
+#endif
+
+int openmc_get_keff(double* k_combined)
+{
+  k_combined[0] = 0.0;
+  k_combined[1] = 0.0;
+
+  // Special case for n <=3. Notice that at the end,
+  // there is a N-3 term in a denominator.
+  if (simulation::n_realizations <= 3) {
+    k_combined[0] = simulation::keff;
+    k_combined[1] = simulation::keff_std;
+    if (simulation::n_realizations <=1) {
+      k_combined[1] = std::numeric_limits<double>::infinity();
+    }
+    return 0;
+  }
+
+  // Initialize variables
+  int64_t n = simulation::n_realizations;
+
+  // Copy estimates of k-effective and its variance (not variance of the mean)
+  const auto& gt = simulation::global_tallies;
+
+  std::array<double, 3> kv {};
+  xt::xtensor<double, 2> cov = xt::zeros<double>({3, 3});
+  kv[0] = gt(K_COLLISION, RESULT_SUM) / n;
+  kv[1] = gt(K_ABSORPTION, RESULT_SUM) / n;
+  kv[2] = gt(K_TRACKLENGTH, RESULT_SUM) / n;
+  cov(0, 0) = (gt(K_COLLISION, RESULT_SUM_SQ) - n*kv[0]*kv[0]) / (n - 1);
+  cov(1, 1) = (gt(K_ABSORPTION, RESULT_SUM_SQ) - n*kv[1]*kv[1]) / (n - 1);
+  cov(2, 2) = (gt(K_TRACKLENGTH, RESULT_SUM_SQ) - n*kv[2]*kv[2]) / (n - 1);
+
+  // Calculate covariances based on sums with Bessel's correction
+  cov(0, 1) = (simulation::k_col_abs - n * kv[0] * kv[1]) / (n - 1);
+  cov(0, 2) = (simulation::k_col_tra - n * kv[0] * kv[2]) / (n - 1);
+  cov(1, 2) = (simulation::k_abs_tra - n * kv[1] * kv[2]) / (n - 1);
+  cov(1, 0) = cov(0, 1);
+  cov(2, 0) = cov(0, 2);
+  cov(2, 1) = cov(1, 2);
+
+  // Check to see if two estimators are the same; this is guaranteed to happen
+  // in MG-mode with survival biasing when the collision and absorption
+  // estimators are the same, but can theoretically happen at anytime.
+  // If it does, the standard estimators will produce floating-point
+  // exceptions and an expression specifically derived for the combination of
+  // two estimators (vice three) should be used instead.
+
+  // First we will identify if there are any matching estimators
+  int i, j, k;
+  if ((std::abs(kv[0] - kv[1]) / kv[0] < FP_REL_PRECISION) &&
+      (std::abs(cov(0, 0) - cov(1, 1)) / cov(0, 0) < FP_REL_PRECISION)) {
+    // 0 and 1 match, so only use 0 and 2 in our comparisons
+    i = 0;
+    j = 2;
+
+  } else if ((std::abs(kv[0] - kv[2]) / kv[0] < FP_REL_PRECISION) &&
+             (std::abs(cov(0, 0) - cov(2, 2)) / cov(0, 0) < FP_REL_PRECISION)) {
+    // 0 and 2 match, so only use 0 and 1 in our comparisons
+    i = 0;
+    j = 1;
+
+  } else if ((std::abs(kv[1] - kv[2]) / kv[1] < FP_REL_PRECISION) &&
+             (std::abs(cov(1, 1) - cov(2, 2)) / cov(1, 1) < FP_REL_PRECISION)) {
+    // 1 and 2 match, so only use 0 and 1 in our comparisons
+    i = 0;
+    j = 1;
+
+  } else {
+    // No two estimators match, so set i to -1 and this will be the indicator
+    // to use all three estimators.
+    i = -1;
```

Comment:
Optional / super nit: It's a small thing, but I actually think having a boolean indicator for this state rather than co-opting `i` would make this cleaner.  I have a small fear that setting `i` to this value could result in silent, hard-to-debug issues if this code is later changed.  In addition, I typically prefer to not co-opt values for these cases unless necessary through hardware, memory, or API constraints.